lib/api/tests/export.rs

```diff
@@ -0,0 +1,264 @@
+            assert_eq!(table.is_strong_instance_ref(), Some(true));
+        }
+
+        let f: NativeFunc<(), ()> = instance.exports.get_native_function("call_host_fn")?;
+        f.call()?;
+        f
+    };
+    f.call()?;
+
+    Ok(())
+}
+
+#[test]
+fn strong_weak_behavior_works_function() -> Result<()> {
+    #[derive(Clone, WasmerEnv, Default)]
+    struct FunctionEnv {
+        #[wasmer(export)]
+        call_host_fn: LazyInit<Function>,
+    }
+
+    let host_fn = |env: &FunctionEnv| {
+        let function = env.call_host_fn_ref().unwrap();
+        assert_eq!(function.is_strong_instance_ref(), Some(false));
+        let function_clone = function.clone();
+        assert_eq!(function_clone.is_strong_instance_ref(), Some(true));
+        assert_eq!(function.is_strong_instance_ref(), Some(false));
+    };
+
+    let f: NativeFunc<(), ()> = {
+        let store = Store::default();
+        let module = Module::new(&store, FUNCTION_WAT)?;
+        let env = FunctionEnv::default();
+
+        let instance = Instance::new(
+            &module,
+            &imports! {
+                "env" => {
+                    "host_fn" => Function::new_native_with_env(&store, env, host_fn)
+                }
+            },
+        )?;
+
+        {
+            let function = instance.exports.get_function("call_host_fn")?;
+            assert_eq!(function.is_strong_instance_ref(), Some(true));
+        }
+
+        let f: NativeFunc<(), ()> = instance.exports.get_native_function("call_host_fn")?;
+        f.call()?;
+        f
+    };
+    f.call()?;
+
+    Ok(())
+}
+
+#[test]
+fn strong_weak_behavior_works_native_function() -> Result<()> {
+    #[derive(Clone, WasmerEnv, Default)]
+    struct FunctionEnv {
+        #[wasmer(export)]
+        call_host_fn: LazyInit<NativeFunc<(), ()>>,
+    }
+
+    let host_fn = |env: &FunctionEnv| {
+        let function = env.call_host_fn_ref().unwrap();
+        assert_eq!(function.is_strong_instance_ref(), Some(false));
+        let function_clone = function.clone();
+        assert_eq!(function_clone.is_strong_instance_ref(), Some(true));
+        assert_eq!(function.is_strong_instance_ref(), Some(false));
+    };
+
+    let f: NativeFunc<(), ()> = {
+        let store = Store::default();
+        let module = Module::new(&store, FUNCTION_WAT)?;
+        let env = FunctionEnv::default();
+
+        let instance = Instance::new(
+            &module,
+            &imports! {
+                "env" => {
+                    "host_fn" => Function::new_native_with_env(&store, env, host_fn)
+                }
+            },
+        )?;
+
+        {
+            let function: NativeFunc<(), ()> =
+                instance.exports.get_native_function("call_host_fn")?;
+            assert_eq!(function.is_strong_instance_ref(), Some(true));
+        }
+
+        let f: NativeFunc<(), ()> = instance.exports.get_native_function("call_host_fn")?;
+        f.call()?;
+        f
+    };
+    f.call()?;
+
+    Ok(())
+}
```

Comment:
We are testing the strong/weak behavior directly (`is_strong_instance_ref`) which is great.

But can we have an extra test that verifies the behavior is fixed without checking for strong/weak fields? (that means, without assuming the design the API)

Reply:
It's not exactly easy to test because it's a memory leak test, it is possible to run some logic in a loop and if it crashes due to OOM then it triggers. My concern about that is that changes to the system (like how much virtual / real memory is available) may change when/if that test passes / fails.